c10/core/DispatchKeySet.h

```diff
@@ -107,65 +317,104 @@ class DispatchKeySet final {
-    // singleton constructor to shift from the right, we can get rid of the
-    // subtraction here.  It's modestly more complicated to get right so I
-    // didn't do it for now.
-    return static_cast<DispatchKey>(64 - llvm::countLeadingZeros(repr_));
+    auto functionality_k = highestFunctionalityKey();
+    if (isPerBackendFunctionalityKey(functionality_k)) {
+      return toRuntimePerBackendFunctionalityKey(functionality_k, highestBackendKey());
+    }
+    return functionality_k;
+  }
+
+  // Returns the index of the most-significant bit in the keyset.
+  // This is used to as part of the calculation into the operator table to get:
+  // - the highest "functionality" bit in the keyset.
+  // - the highest "backend" bit in the keyset.
+  uint8_t indexOfHighestBit() const {
+    return 64 - llvm::countLeadingZeros(repr_);
   }
 
-  DispatchKey highestPriorityBackendTypeId() const {
-    return (*this &
-            ((1ULL << static_cast<uint8_t>(DispatchKey::EndOfBackendKeys)) - 1))
-        .highestPriorityTypeId();
+  // returns the index in the operator table of highest priority key in the the keyset
+  // Note that we could in theory implement this using highestPriorityTypeId(),
+  // but this code is very hotpath and we can do it faster without it.
+  uint64_t getDispatchTableIndexForDispatchKeySet() const {
+    auto functionality_idx =
+        DispatchKeySet(repr_ >> num_backends)
+           .indexOfHighestBit();
+    auto offset_and_mask = offsetsAndMasks()[functionality_idx];
+    // Mask the functionality bits out first, then right-shift by 1.
+    // right-shifting by 1 because everything is zero-indexed.
+    // E.g. 000001 (CPU) should give us an offset of 0, 000010 (CUDA) should give us an offset of 1, etc.
+    auto backend_idx = DispatchKeySet((repr_ & offset_and_mask.mask) >> 1).indexOfHighestBit();
+    return offset_and_mask.offset + backend_idx;
   }
 
  private:
   constexpr DispatchKeySet(uint64_t repr) : repr_(repr) {}
   uint64_t repr_ = 0;
 
  public:
-  // STL iterator for DispatchKeySet. Iterates through all DispatchKeys in the
-  // set. The iterator is only invalidated by the destruction of the underlying
+  // STL iterator for DispatchKeySet. Iterates through all runtime DispatchKeys in
+  // the set. The iterator is only invalidated by the destruction of the underlying
   // DispatchKeySet as the iterator stores a pointer to the raw representation
   // of the DispatchKeySet.
+  // Note: When we encounter a per-backend functionality (e.g. Dense or Sparse),
+  // we will iterate through EVERY backend in the keyset, for that functionality.
+  // For example, if the next functionality key to iterate over is Autograd,
+  // and the backend bits in the keyset correspond to [BackendBit::CPUBit, DispatchKey::CUDABit],
+  // then the next two keys we return will be DispatchKey::AutogradCPU, DispatchKey::AutogradCUDA
+  // (CPU first because it has lower precedence than CUDA in DispatchKey.h).
```

Comment:
That's cuda first and then cpu right?

Reply:
The DispatchKeySet iterator actually iterates from low -> high in terms of precedence - the opposite from our "dispatch decision" ordering.

I'm not actually sure if there's a meaningful reason for it, but the iterator is only really used in a few places at registration time, where the ordering isn't important. I figured I'd keep the functionality we already have.

Reply:
I think it would be nice to mention this here